From 3637c4fe313d466ee4275f32fb9c5c6480d32ae2 Mon Sep 17 00:00:00 2001
From: Karthik D A <karthikda@maxlinear.com>
Date: Mon, 1 Sep 2025 02:02:07 +0800
Subject: [PATCH] WLANRTSYS-93718 iwlwav-hostapd-ng MSCS validation

Description: After the implementation of the MSCS changes, MSCS QM TCs failed.

Fix: Include handling of the MSCS descriptor in assoc response and implementation of the unsolicited MSCS resp.

Test: Tested and verified that the MSCS TC 4.4.1_5G, 4.4.2_24G/5G and 4.4.3_24/5G passed successfully.

Fix Up: WLANRTSYS-92508 iwlwav-hostapd-ng: MSCS Support

Signed-off-by: Karthik D A <karthikda@maxlinear.com>
---
 hostapd/mxl_ctrl_iface.c       | 27 +++++++++++++++++++++++++++
 hostapd/mxl_hostapd_cli.c      | 18 ++++++++++++++++++
 hostapd/mxl_hostapd_cli.h      |  4 ++++
 src/ap/mxl_ieee802_11_shared.c |  2 --
 src/ap/mxl_wmm.c               | 29 +++++++++++++++--------------
 src/ap/mxl_wmm.h               |  3 ++-
 6 files changed, 66 insertions(+), 17 deletions(-)

diff --git a/hostapd/mxl_ctrl_iface.c b/hostapd/mxl_ctrl_iface.c
index f4d4e93fc..7a9def3f0 100644
--- a/hostapd/mxl_ctrl_iface.c
+++ b/hostapd/mxl_ctrl_iface.c
@@ -35,6 +35,7 @@
 #ifdef CONFIG_AFC
 #include "ap/mxl_afc.h"
 #endif /* CONFIG_AFC */
+#include "ap/mxl_wmm.h"
 
 struct hostapd_data *mxl_get_bss_index(const char **cmd, struct hostapd_data *hapd, bool ctrl_iface, bool delim)
 {
@@ -3162,6 +3163,29 @@ static int mxl_hostapd_ctrl_iface_get_mscs_enable(struct hostapd_data *hapd, cha
 
 	return res;
 }
+
+static int mxl_hostapd_ctrl_iface_mscs_unsolicited_rem(struct hostapd_data *hapd, const char *cmd, char *reply, size_t reply_size)
+{
+	u8 addr[ETH_ALEN];
+	int status;
+	int res = 0;
+
+	if (hwaddr_aton2(cmd, addr) < 0) {
+		wpa_printf(MSG_ERROR, "MSCS: Invalid MAC '%s'", cmd);
+		return -1;
+	}
+
+	status = mxl_mscs_remove_req(hapd, addr, 0);
+	mxl_send_mscs_resp(hapd, addr, status, 0);
+
+	res = sprintf_s(reply, reply_size, "%d\n", status);
+	if (res <= 0) {
+		wpa_printf(MSG_ERROR, "%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+
+	return res;
+}
 #endif /* CONFIG_MSCS_MXL */
 
 int mxl_hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
@@ -3400,6 +3424,9 @@ int mxl_hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 	} else if (os_strncmp(buf, "GET_MSCS_ENABLE", 15) == 0) {
 		if (mxl_hostapd_ctrl_iface_get_mscs_enable(hapd, reply, reply_size) < 0)
 			*reply_len = -1;
+	} else if (os_strncmp(buf, "MSCS_REM_REQ ", sizeof("MSCS_REM_REQ ") - 1) == 0) {
+		if (mxl_hostapd_ctrl_iface_mscs_unsolicited_rem(hapd, buf + sizeof("MSCS_REM_REQ ") -1, reply, reply_size) < 0)
+			*reply_len = -1;
 #endif /* CONFIG_MSCS */
 	} else {
 		return 0;
diff --git a/hostapd/mxl_hostapd_cli.c b/hostapd/mxl_hostapd_cli.c
index 100e2b852..c30003995 100644
--- a/hostapd/mxl_hostapd_cli.c
+++ b/hostapd/mxl_hostapd_cli.c
@@ -569,3 +569,21 @@ int mxl_hostapd_cli_cmd_get_mscs_enable(struct wpa_ctrl *ctrl, int argc, char *a
 {
 	return mxl_wpa_ctrl_command(ctrl, "GET_MSCS_ENABLE");
 }
+
+int mxl_hostapd_cli_cmd_mscs_rem(struct wpa_ctrl *ctrl, int argc, char *argv[])
+{
+   char cmd[256];
+	int res;
+
+	if (argc != 1) {
+		printf("Usage: mscs_remove <STA-MAC>\n");
+		return -1;
+	}
+
+	res = sprintf_s(cmd, sizeof(cmd), "MSCS_REM_REQ %s", argv[0]);
+	if (res <= 0) {
+		printf("%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+	return mxl_wpa_ctrl_command(ctrl, cmd);
+}
diff --git a/hostapd/mxl_hostapd_cli.h b/hostapd/mxl_hostapd_cli.h
index 9e8e1da72..5892ce4b3 100644
--- a/hostapd/mxl_hostapd_cli.h
+++ b/hostapd/mxl_hostapd_cli.h
@@ -111,6 +111,8 @@ int mxl_wpa_ctrl_command(struct wpa_ctrl *ctrl, const char *cmd);
 	  " = Get scs enable/disable status"}, \
 	{ "get_mscs_enable", mxl_hostapd_cli_cmd_get_mscs_enable, NULL, \
 	  " = Get mscs enable/disable status"}, \
+	{ "mscs_remove", mxl_hostapd_cli_cmd_mscs_rem, NULL, \
+	  " = Unsolicited mscs rem req"}, \
 
 #ifdef CONFIG_ACS
 #define MXL_CLI_COMMANDS_ACS \
@@ -193,6 +195,8 @@ int mxl_hostapd_cli_cmd_get_assocresp_elements(struct wpa_ctrl *ctrl, int argc,
 int mxl_hostapd_cli_cmd_bss_reconf(struct wpa_ctrl *ctrl, int argc, char *argv[]);
 int mxl_hostapd_cli_cmd_get_scs_enable(struct wpa_ctrl *ctrl, int argc, char *argv[]);
 int mxl_hostapd_cli_cmd_get_mscs_enable(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_mscs_rem(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+
 
 
 #endif /* MXL_HOSTAPD_CLI_H */
diff --git a/src/ap/mxl_ieee802_11_shared.c b/src/ap/mxl_ieee802_11_shared.c
index deaa50cb4..2394e9b64 100644
--- a/src/ap/mxl_ieee802_11_shared.c
+++ b/src/ap/mxl_ieee802_11_shared.c
@@ -145,8 +145,6 @@ u8 * mxl_hostapd_eid_mscs(struct hostapd_data *hapd, u8 *eid, size_t len, u16 st
 	u8 elen = MSCS_IE_ASSOC_RESP_LEN;
 
 	if (!(hapd->conf->mxl_bss_conf.mscs_enable && sta &&
-				sta->mxl_sta_info.sta_mscs_ie.mscs_descriptor &&
-				sta->mxl_sta_info.sta_mscs_ie.mscs_descriptor_len &&
 				sta->mxl_sta_info.sta_mscs_ie.mscs_active))
 		return eid;
 
diff --git a/src/ap/mxl_wmm.c b/src/ap/mxl_wmm.c
index 6ca9936e6..9df5fe8a6 100644
--- a/src/ap/mxl_wmm.c
+++ b/src/ap/mxl_wmm.c
@@ -612,7 +612,6 @@ void mxl_handle_scs_req(struct hostapd_data *hapd,
 #ifdef CONFIG_MSCS_MXL
 int mxl_mscs_parse_tclas_mask_elems(const u8 *pos, struct mxl_mscs_add_req *mscs_add_req)
 {
-	u8 class_type;
 
 	if (*pos++ != WLAN_EID_EXT_TCLAS_MASK)
 		return WLAN_STATUS_UNSPECIFIED_FAILURE;
@@ -697,14 +696,14 @@ int mxl_mscs_add_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgm
 	return WLAN_STATUS_SUCCESS;
 }
 
-static void mxl_send_mscs_resp(struct hostapd_data *hapd,
-			const struct ieee80211_mgmt *mgmt, int status)
+void mxl_send_mscs_resp(struct hostapd_data *hapd,
+			const u8 *addr, int status, u8 dialog_token)
 {
 	u8 buf[256];
 	struct ieee80211_mgmt *mgmt_resp = (struct ieee80211_mgmt *) buf;
 	struct sta_info *sta;
 
-	sta = ap_get_sta(hapd, mgmt->sa);
+	sta = ap_get_sta(hapd, addr);
 	if (sta == NULL) {
 		wpa_printf(MSG_ERROR, "MSCS: mxl_send_scs_resp - station not connected");
 		return;
@@ -713,13 +712,13 @@ static void mxl_send_mscs_resp(struct hostapd_data *hapd,
 	wpa_printf(MSG_DEBUG, "handle_mscs_resp");
 	os_memset(buf, 0, sizeof(buf));
 	mgmt_resp->frame_control = IEEE80211_FC(WLAN_FC_TYPE_MGMT, WLAN_FC_STYPE_ACTION);
-	os_memcpy(mgmt_resp->da, mgmt->sa, ETH_ALEN);
+	os_memcpy(mgmt_resp->da, addr, ETH_ALEN);
 	os_memcpy(mgmt_resp->sa, hapd->own_addr, ETH_ALEN);
 	os_memcpy(mgmt_resp->bssid, hapd->own_addr, ETH_ALEN);
 
 	mgmt_resp->u.action.category = WLAN_ACTION_ROBUST_AV_STREAMING;
 	mgmt_resp->u.action.u.mscs_action_resp.action = ROBUST_AV_MSCS_RESP;
-	mgmt_resp->u.action.u.mscs_action_resp.dialog_token = mgmt->u.action.u.mscs_action_req.dialog_token;
+	mgmt_resp->u.action.u.mscs_action_resp.dialog_token = dialog_token;
 	mgmt_resp->u.action.u.mscs_action_resp.status = status;
 	if (hostapd_drv_send_mlme(hapd, mgmt_resp, IEEE80211_HDRLEN + sizeof(mgmt_resp->u.action.u.mscs_action_resp) +
 										sizeof(mgmt_resp->u.action.category), 0, NULL, 0, 0) < 0)
@@ -727,12 +726,12 @@ static void mxl_send_mscs_resp(struct hostapd_data *hapd,
 }
 
 int mxl_mscs_remove_req(struct hostapd_data *hapd,
-		const struct ieee80211_mgmt *mgmt, size_t len)
+		const u8 *addr, size_t len)
  {
 	struct sta_info *sta;
 	u16 aid;
 
-	sta = ap_get_sta(hapd, mgmt->sa);
+	sta = ap_get_sta(hapd, addr);
 	if (sta == NULL) {
 		wpa_printf(MSG_ERROR, "MSCS: mscs_rem_req - station not connected");
 		return WLAN_STATUS_UNSPECIFIED_FAILURE;
@@ -760,7 +759,7 @@ void mxl_handle_mscs_req(struct hostapd_data *hapd,
 			const struct ieee80211_mgmt *mgmt, size_t len)
 {
 	int status;
-
+	u8 dialog_token;
 	const u8 *pos = mgmt->u.action.u.mscs_action_req.variable;
 	u8 req_type, ext_eid, eid, elen;
 
@@ -786,21 +785,23 @@ void mxl_handle_mscs_req(struct hostapd_data *hapd,
 	req_type = *pos++;
 	elen --; /* Subtract req_type */
 
+	dialog_token = mgmt->u.action.u.mscs_action_req.dialog_token;
+
 	switch(req_type)
 	{
 		case MSCS_REQ_ADD:
 			status = mxl_mscs_add_req(hapd, mgmt, pos, elen);
-			mxl_send_mscs_resp(hapd, mgmt, status);
+			mxl_send_mscs_resp(hapd, mgmt->sa, status, dialog_token);
 			break;
 		case MSCS_REQ_REMOVE:
-			status = mxl_mscs_remove_req(hapd, mgmt, elen);
-			mxl_send_mscs_resp(hapd, mgmt, status);
+			status = mxl_mscs_remove_req(hapd, mgmt->sa, elen);
+			mxl_send_mscs_resp(hapd, mgmt->sa, status, dialog_token);
 			break;
 		case MSCS_REQ_CHANGE:
-			status = mxl_mscs_remove_req(hapd, mgmt, elen);
+			status = mxl_mscs_remove_req(hapd, mgmt->sa, elen);
 			if (status == WLAN_STATUS_TCLAS_PROCESSING_TERMINATED)
 				status = mxl_mscs_add_req(hapd, mgmt, pos, elen);
-			mxl_send_mscs_resp(hapd, mgmt, status);
+			mxl_send_mscs_resp(hapd, mgmt->sa, status, dialog_token);
 			break;
 		default:
 			hostapd_logger(hapd, mgmt->sa, HOSTAPD_MODULE_IEEE80211,
diff --git a/src/ap/mxl_wmm.h b/src/ap/mxl_wmm.h
index f7c5a2d6f..81d177a35 100644
--- a/src/ap/mxl_wmm.h
+++ b/src/ap/mxl_wmm.h
@@ -39,7 +39,8 @@ int mxl_scs_parse_tclas_elems(const u8 *pos, struct mxl_scs_add_req *scs_add_req
 #ifdef CONFIG_MSCS_MXL
 void mxl_handle_mscs_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, size_t len);
 int mxl_mscs_add_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, const u8 *pos, size_t len);
-int mxl_mscs_remove_req(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt, size_t len);
+int mxl_mscs_remove_req(struct hostapd_data *hapd, const u8 *addr, size_t len);
 int mxl_mscs_parse_tclas_mask_elems(const u8 *pos, struct mxl_mscs_add_req *mscs_add_req);
+void mxl_send_mscs_resp(struct hostapd_data *hapd, const u8 *addr, int status, u8 dialog_token);
 #endif /* CONFIG_MSCS_MXL */
 #endif /* MXL_WME_H */
-- 
2.43.0

